File: 2025/07 July/08 Next element with greater frequency/main.cpp
// Link: https://www.geeksforgeeks.org/problems/next-element-with-greater-frequency--170637/1

#include <iostream>
#include <vector>
#include <stack>
#include <map>
using namespace std;

class Solution
{
public:
    vector<int> findGreater(vector<int> &arr)
    {
        int n = arr.size();
        map<int, int> freq;

        for (auto it : arr)
        {
            freq[it]++;
        }

        vector<int> res(n, -1);
        stack<int> s;

        for (int i = 0; i < n; i++)
        {
            while (!s.empty() && freq[arr[i]] > freq[arr[s.top()]])
            {
                res[s.top()] = arr[i];
                s.pop();
            }
            s.push(i);
        }

        return res;
    }
};

int main()
{
    Solution sol;
    vector<int> arr = {2, 1, 1, 3, 2, 1};
    vector<int> result = sol.findGreater(arr);
    for (int i : result)
    {
        cout << i << " ";
    }
    cout << endl;

    arr = {5, 1, 5, 6, 6};
    result = sol.findGreater(arr);
    for (int i : result)
    {
        cout << i << " ";
    }
    cout << endl;

    return 0;
}

File: 2025/07 July/08 Next element with greater frequency/main.py
# Link: https://www.geeksforgeeks.org/problems/next-element-with-greater-frequency--170637/1


class Solution:
    def findGreater(self, arr):
        n = len(arr)
        freq = {}

        for num in arr:
            freq[num] = freq.get(num, 0) + 1

        res = [-1] * n
        s = []

        for i in range(n):
            while s and freq[arr[i]] > freq[arr[s[-1]]]:
                res[s.pop()] = arr[i]
            s.append(i)

        return res


if __name__ == "__main__":
    sol = Solution()
    arr = [2, 1, 1, 3, 2, 1]
    result = sol.findGreater(arr)
    print(result)

    arr = [5, 1, 5, 6, 6]
    result = sol.findGreater(arr)
    print(result)
